subprojects/model-core/src/main/java/org/gradle/internal/instantiation/generator/AsmBackedClassGenerator.java

```diff
@@ -1137,7 +1152,17 @@ public void applyServiceInjectionToGetter(PropertyMetadata property, @Nullable f
 
                 // (<type>)<service>
                 _CHECKCAST(serviceType);
-            }});
+            }};
+
+            BytecodeFragment getterEpilogue = methodVisitor -> new LocalMethodVisitorScope(methodVisitor) {{
+                // InstrumentedExecutionAccess.injectedServiceAccessed(<services>, <service-type>, <this-type-name>)
+                putServiceRegistryOnStack();
+                _LDC(serviceType);
+                _LDC(type.getName());
+                _INVOKESTATIC(INSTRUMENTED_EXECUTION_ACCESS_TYPE, "injectedServiceAccessed", getMethodDescriptor(VOID_TYPE, SERVICE_LOOKUP_TYPE, CLASS_TYPE, STRING_TYPE));
```

Comment:
I wonder if we can only instrument "interesting" getters to avoid extra costs on every service access.

Reply:
We definitely could, my only concern is that `AsmBackedClassGenerator` will have to know about what services are allowed, what is CC specific information.